File: projects/grind-75/python/string_lc_questions/longest_substring_without_repeating_characters.py
class Solution:
    @classmethod
    def lengthOfLongestSubstring(self, s: str) -> int:
        c_index = {}
        left = 0
        res = 0
        for right, c in enumerate(s):
            if c in c_index and left <= c_index[c]:
                left = c_index[c] + 1
            c_index[c] = right
            res = max(res, right - left + 1)
        return res


"""
Time Complexity - O(n)
Space Complexity - O(n)
"""

File: projects/grind-75/typescript/string/longest-substring-without-repeating-characters.ts
export default function lengthOfLongestSubstring(s: string): number {
  const cIndex: Record<string, number> = {};
  let left = 0;
  let res = 0;

  for (let right = 0; right < s.length; right++) {
    const c = s[right];
    if (c in cIndex && left <= cIndex[c]) {
      left = cIndex[c] + 1;
    }
    cIndex[c] = right;
    res = Math.max(res, right - left + 1);
  }

  return res;
}
